### PR TITLE
Mongo refactoring to support functions in LHS of filters

### PR DESCRIPTION
**Pull request by suresh-prakash**
## Description
1. Introduction of new classes for "$addFields" support in MongoDB
2. Override toString() for the ease of debugging

### Tests:
Nothing to test since new classes have been added, yet not integrated yet.

### Checklist:
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: document-store/src/test/java/org/hypertrace/core/documentstore/query/QueryBuilderTest.java [hunk 1]
```diff
@@ -23,6 +45,253 @@ void setUp() {
...
+                            IdentifierExpression.of("percentage"), GT, ConstantExpression.of(90)))
+                    .operator(AND)
+                    .operand(
+                        RelationalExpression.of(
+                            IdentifierExpression.of("class"), EQ, ConstantExpression.of("XII")))
+                    .build())
+            .build();
+
+    assertEquals(
+        "SELECT * FROM <implicit_collection> WHERE (`percentage` > 90) AND (`class` = 'XII')",
```
Comment (kotharironak) [inline document-store/src/test/java/org/hypertrace/core/documentstore/query/QueryBuilderTest.java:81]: Why are we using this ` one. Can we use a simple quote? If needed can be escaped

Comment (suresh-prakash) [reply document-store/src/test/java/org/hypertrace/core/documentstore/query/QueryBuilderTest.java:81]: That is to clearly indicate that it is an Identifier and not a constant.
Constants strings are surrounded with quotes.

Note: Wrapping column names under back ticks (`) is common in some SQL platforms like [MySQL](https://www.tutorialspoint.com/how-to-select-a-column-name-with-spaces-in-mysql#:~:text=To%20select%20a%20column%20name%20with%20spaces%2C%20use%20the%20back,below%20the%20tilde%20operator%20(%20~).)

File: document-store/src/main/java/org/hypertrace/core/documentstore/expression/operators/RelationalOperator.java [hunk 2]
```diff
@@ -1,16 +1,27 @@
...
+  GT(">"),
+  LT("<"),
+  GTE(">="),
+  LTE("<="),
+  IN("IN"),
+  CONTAINS("CONTAINS"),
+  EXISTS("EXISTS"),
+  NOT_EXISTS("NOT EXISTS"),
+  LIKE("~"),
+  NOT_IN("NOT IN"),
```
Comment (kotharironak) [inline document-store/src/main/java/org/hypertrace/core/documentstore/expression/operators/RelationalOperator.java:18]:
```suggestion
  NOT_IN("NOT IN");
```

Comment (kotharironak) [reply document-store/src/main/java/org/hypertrace/core/documentstore/expression/operators/RelationalOperator.java:18]: nit: if you do this remove the below the line.

File: document-store/src/main/java/org/hypertrace/core/documentstore/mongo/parser/MongoAddFieldsParser.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        .flatMap(Set::stream)
+        .collect(
+            collectingAndThen(
+                toUnmodifiableMap(Entry::getKey, Entry::getValue, (oldVal, newVal) -> oldVal),
+                BasicDBObject::new));
+  }
+
+  @SuppressWarnings("unchecked")
+  private static Map<String, Object> getNewFieldsForFiltering(final Query query) {
+    final MongoFilterFieldAddingParser filterFieldAddingParser = new MongoFilterFieldAddingParser();
```
Comment (kotharironak) [inline document-store/src/main/java/org/hypertrace/core/documentstore/mongo/parser/MongoAddFieldsParser.java:38]: nit: `MongoFilterFieldAddingParser` -> `MongoAddFieldsFilterParser` or `MongoFilterAddFieldsParser` for ease of readability.

Comment (kotharironak) [reply document-store/src/main/java/org/hypertrace/core/documentstore/mongo/parser/MongoAddFieldsParser.java:38]: if change above, also change var - `filterFieldAddingParser` to appropriate.

**Review by kotharironak — COMMENTED** [review]
lgtm, have nit comments. Let me know if you are addressing else we can move forward.

File: document-store/src/main/java/org/hypertrace/core/documentstore/expression/operators/RelationalOperator.java (hunk 2 above)
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/expression/operators/RelationalOperator.java:18]: Prefer to keep semicolon in new line for allowing operator extendibility without affecting the git blame of current line.

**Review by kotharironak — APPROVED** [review]

**Pull request merged**